crates/sui-core/src/node_sync/node_follower.rs

```diff
@@ -233,51 +262,41 @@ where
             pending_txes: Waiter::new(),
         }
     }
+}
 
-    fn start(self, receiver: mpsc::Receiver<DigestsMessage>) -> JoinHandle<()> {
-        let state = Arc::new(self);
-        tokio::spawn(async move {
-            state
-                .handle_stream(SyncMode::Follow, ReceiverStream::new(receiver))
-                .await
-        })
-    }
-
-    pub async fn sync_checkpoint(self, checkpoint_contents: &CheckpointContents) -> SuiResult {
-        let stream = tokio_stream::iter(
-            checkpoint_contents
-                .transactions
-                .iter()
-                .map(DigestsMessage::new_for_ckpt),
-        );
-        let state = Arc::new(self);
-        state.handle_stream(SyncMode::Checkpoint, stream).await;
-        Ok(())
+impl<A> NodeSyncState<A>
+where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+{
+    fn start(self: Arc<Self>, receiver: mpsc::Receiver<DigestsMessage>) -> JoinHandle<()> {
+        let state = self;
+        tokio::spawn(async move { state.handle_stream(ReceiverStream::new(receiver)).await })
     }
 
-    async fn handle_stream(
-        self: Arc<Self>,
-        mode: SyncMode,
-        stream: impl Stream<Item = DigestsMessage>,
-    ) {
+    async fn handle_stream(self: Arc<Self>, stream: impl Stream<Item = DigestsMessage>) {
         // this pattern for limiting concurrency is from
         // https://github.com/tokio-rs/tokio/discussions/2648
         let limit = Arc::new(Semaphore::new(MAX_NODE_SYNC_CONCURRENCY));
         let mut stream = Box::pin(stream);
 
-        while let Some(DigestsMessage { digests, peer, tx }) = stream.next().await {
+        while let Some(DigestsMessage { sync_arg, peer, tx }) = stream.next().await {
```

Comment:
heads up: efficient parallel execution can be best implemented if you pass in blocks of transactions rather than passing the transactions in one by one.

Reply:
What's your argument here for efficiency? I can imagine some theoretical benefits (e.g. cache coherency, less synchronization overhead in the channels) that might be obtained from better batching, but I wouldn't expect that to have a noticeable effect here.

Also - keep in mind that even if we pass in TXes in blocks, we are going to want to farm the execution out to multiple tasks for execution parallelism anyway. (This is a TODO right now - i'm waiting until I have a working devnet to do that so I can measure how much of a speedup it is).